### ISSUE TITLE
Proposal: Enhance custom inspect API

### ISSUE THREAD

**Issue by kitsonk** [denoland/deno#8099]
*Edit: the current proposal is here https://github.com/denoland/deno/issues/8099#issuecomment-757434256*

---

Probably should have thought of this when we did `Deno.customInspect` but we didn't.  As our console output has evolved, it makes it really hard to implement the same rich output.  For example if you have the following code:

```js
const h = new Headers();
h.append("key", "value");
console.log(h);
```

You would get something like the following output:

<img width="157" alt="a non coloured output string" src="https://user-images.githubusercontent.com/1282577/97069044-d6944700-1618-11eb-87f7-37365013b3ac.png">

Where as if you did:

```js
console.log({ key: "value" });
```

You would get:

<img width="115" alt="a coloured output string" src="https://user-images.githubusercontent.com/1282577/97069063-017e9b00-1619-11eb-940d-35428f56c4d9.png">

Which takes advantage of all the other benefits of the output algorithm.

Therefore if we had another symbol that returned a more structured set of information, we could use the same algorithms to colourise the output and otherwise pretty print it.  I would propose that we create `Symbol("Deno.richCustomInspect")` which would require a return of a tuple value of two strings.  The first string would be the _name_ for the inspector and the second value would be a JSON string, which would be parsed by the inspector.

So for example if I wanted a rich output for `Headers`, I would do something like this:

```js
class Headers {
  [Symbol.for("Deno.richCustomInspect]() {
    const output = {};
    for (const [key, value] of this[headersData]) {
      output[key] = value;
    }
    return [ "Headers", JSON.stringify(output) ];
  }
}
```

This would allow complex modelling of the internals, but ensure serialisation to a consistent format that can be handled by the inspector.

**Comment by caspervonb** [comment]
Instead of adding more symbols, would be simpler if the results of calling custom inspect were just formatted with the object format specifier instead (raw strings are left alone with that).

**Comment by kitsonk** [comment]
@caspervonb I assume you are suggesting something like this PR for `Event`? https://github.com/denoland/deno/pull/8103/files

It hadn't occurred to me that should be the way until I saw the PR.  We could easily clean up `Headers` with something like that.

**Comment by caspervonb** [comment]
Pretty much!

Side note tho, customInspect signature should probably be given `depth:number` and `options: InspectOptions` so the formatter used can respects those settings.

**Comment by kitsonk** [comment]
@caspervonb if it was just passed `options`, it includes the depth already, and would be implicit that one would want to likely decrement that when passing it on.  What about something like:

```ts
type CustomInspect = (inspect: (value: unknown, options?: InspectOptions) => string, options: InspectOptions) => string;
```

That way the implementor wouldn't need to worry about having a reference to the `Deno` namespace, which could also be implementing some sort of custom inspection function as well (like sandboxing or logging, etc.).  Sort of like having the formatter passed in Rust.

**Comment by caspervonb** [comment]
> @caspervonb if it was just passed options, it includes the depth already, and would be implicit that one would want to likely decrement that when passing it on. What about something like:

Yeah, if it's already embedded in options then a single optional is the way to go; also doesn't break anything which is a bonus 👍

**Comment by nayeemrmn** [comment]
I'm for the alternative of just relying on `Deno.inspect()` recursions.

> Side note tho, customInspect signature should probably be given `depth:number` and `options: InspectOptions` so the formatter used can respects those settings.

I've opened #9293.

> That way the implementor wouldn't need to worry about having a reference to the Deno namespace

I've opened #8174, #8178... it's a shame that ever happened.

**Comment by caspervonb** [comment]
> @caspervonb if it was just passed `options`, it includes the depth already, and would be implicit that one would want to likely decrement that when passing it on. What about something like:
> 
> ```ts
> type CustomInspect = (inspect: (value: unknown, options?: InspectOptions) => string, options: InspectOptions) => string;
> ```
> 
> That way the implementor wouldn't need to worry about having a reference to the `Deno` namespace, which could also be implementing some sort of custom inspection function as well (like sandboxing or logging, etc.). Sort of like having the formatter passed in Rust.

Right, I think you meant the inspection depth which is in options, but I meant indentation level which is not included at the moment.

To keep the indentation from resetting we need to pass along the indent depth, so the signature has to be `function (indent: Number, options: InspectOptions);`.

But we should probably also pass the console context as-well.

Was implementing this and I noticed that you're passing around Deno.inspect as the first parameter @nayeemrmn?

**Comment by bartlomieju** [comment]
`Symbol.for("Deno.customInspect")` is the solution in 2025
